### PR TITLE
Add ADR 17

### PR DESCRIPTION
**Pull request by srallen**
For #694.

I have two questions from writing this that I still think we should decide upon:
- Should we continue to use https://fe-project.preview.zooniverse.org as the destination sub-domain for the master branch staging deploys? As we add more apps, using `fe-project` doesn't really make sense and won't accurately reflect what's deployed. I believe currently both `fe-project` and `fe-content-pages` are built and deployed on merge to master.
  - Should individual apps be deployed in their own images and process?
- Should we have a staging webpack build for the libraries to not uglify the builds so we can more easily debug using the staging app?

### CODE REVIEW

File: docs/arch/adr-17.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+## Status
+
+Accepted
+
+## Consequences
+
+- The deployment of the front-end next.js apps will match how the back-end apps are managed so we will now be consistent in process
+- We will need to train developers on the new deployment process
+- We do not have a clear set of steps for setting up the as needed branch deployments at this time
+  - We need to do more research into technically how to accomplish the ad hoc branch deploymenets and tear downs. It is not known if Helm can be used for this.
```
Comment (camallen) [inline docs/arch/adr-17.md]:
```suggestion
  - We need to do more research into technically how to accomplish the ad hoc branch deployments and tear downs. It is not known if Helm can be used for this.
```

**Comment by adammcmaster** [comment]
> What's the reason to move it, clarity of the name and what the domain does? fe doesn't tell me much to be honest and what does project mean in this context considering our domain vocabulary?

"project" refers to the fact that it was set up as the staging environment for the project page app specifically.

frontent.preview.zooniverse.org might be a better name for it.

**Comment by rogerhutchings** [comment]
From talking to Cam this morning:

- In practice, we're going to have two kinds of pull request: one that changes a single app (e.g. new widget on project home page), and one that affects multiple apps (e.g. update to the shared component library)
- For PRs on a single app, we'd like to manually deploy it so it can be tested in isolation. On merging to master, that gets deployed to staging. We'd then do integration testing before manually deploying to production.
- For PRs across multiple apps, we'd test it locally before merging to master. Once it's deployed to staging, we'd do integration testing before manually deploying to production.

Please edit if I've missed anything @camallen

**Comment by camallen** [comment]
Yes that's the idea, avoid deploy the complexity when deploying all apps together combined with an apps feature branch. We can test the changes differently depending on what changes they are, app vs shared code (impacts all apps) and run integration tests (automate some behaviour scripts) via staging systems.

> For PRs on a single app, we'd like to manually deploy it so it can be tested in isolation. On merging to master, that gets deployed to staging. We'd then do integration testing before manually deploying to production.

I think if a PR is only impacting the behaviour of a single app ideally we feature branch deploy just this app (automatically if we can) to test in isolation. This would reduce the complexity of the deploy by only considering the local app state.

We can still test app interactions locally via docker and against other current apps states once a merge to master deploys the app to staging systems.

> For PRs across multiple apps, we'd test it locally before merging to master. Once it's deployed to staging, we'd do integration testing before manually deploying to production.

Test locally in docker, issue a PR, test feature branch deploy works as expected and after merge to master run the integration tests to check the apps work together.

**Comment by camallen** [comment]
I think it’s important that folks dev’ing on the app know how this ‘glue’ works to bind the apps together on a consistent domain target.

Perhaps we can add a section about how cloudfront works as well? Some starting copy..

The Cloudfront configs that wrap the current front end (pfe / monorepo) domain deploy URLs and will split traffic to the correct deployed microservice apps depending on the rules setup in cloudfront.

Both the staging and production domains (www.zooniverse.org) will have cloudfront configurations that will match URL traffic against rules setup in cloudfront.

The rules match paths on the URL, i.e. /about/team maps to the correct app via DNS resolution, e.g. about/team maps to fe-content-pages production app that hits the k8 cluster via DNS.

A GET request on URL `www.zooniverse.org/about/team` will route like this via cloudfront
```
Cloudfront -> k8 ingress (DNS) -> a fe-content-pages pod
```

Generally staging and production would have the same behaviour mappings in cloudfront and staging will be a place to test these mapping out before setting up in production. I can envisage times that staging would more paths i.e. when we add apps.

**Comment by srallen** [comment]
> What's the reason to move it, clarity of the name and what the domain does? fe doesn't tell me much to be honest and what does project mean in this context considering our domain vocabulary?

Yes for clarity. 

> Won't the fe-project solely be the classifier component and pages or will it include other pages (apps?) including the project page landing pages, etc?

Since all the apps are deployed together in one go, then if we have a single sub-domain name and cloudfront setup to route to the correct app, then the current sub-domain doesn't accurately reflect what is being loaded necessarily. I like @adammcmaster suggested sub-domain naming. 

> It seems they are via the jenkins & kubernetes files, am i missing something?

What I mean is that if a PR is only changing the project app, then should only the project-app be deployed upon merge? It looks like this was already discussed in the latter comments :) 

> Can we use source mapping for minified code to handle this instead? Dev tools should handle them by default if defined, https://developers.google.com/web/tools/chrome-devtools/javascript/source-maps

The production webpack configs appear to create them, but I should confirm that. I just know that the uglified javascript is not useful when debugging production. It may be just a matter of tweaking configuration to make sure source maps are functioning as expected.

**Comment by srallen** [comment]
@camallen and @rogerhutchings I've updated the ADR text to include more details about the PR process will be and what Cloudfront's role is. Let me know if this is reading clearly.

**Pull request merged**